e2e/src/tests.rs

```diff
@@ -0,0 +1,288 @@
+        server.kill().unwrap();
+        delete_namespace(&client, namespace).await;
+        sleep(Duration::from_secs(10)).await; // Todo: check if we still need this sleep
+    }
+
+    // agent namespace tests
+    #[tokio::test]
+    async fn test_good_agent_namespace() {
+        // creates a new k8s namespace, starts the API server with env:
+        // MIRRORD_AGENT_NAMESPACE=namespace, asserts that the agent job and pod are created
+        // validate data through requests to the API
+        let mut config = Config::infer().await.unwrap();
+        config.accept_invalid_certs = true;
+        let client = Client::try_from(config).unwrap();
+
+        let namespace = "test-namespace-agent-good";
+        create_namespace(&client, namespace).await;
+        let service_url = get_service_url(&client, "default").await.unwrap();
+        let pod_name = get_nginx_pod_name(&client, "default").await.unwrap();
+        let env = HashMap::from([("MIRRORD_AGENT_NAMESPACE", namespace)]);
+        let command = vec!["node", "node-e2e/app.js"];
+        let mut server = start_server(&pod_name, command, env);
+        sleep(Duration::from_secs(5)).await;
+        match server.try_wait() {
+            Ok(Some(status)) => {
+                delete_namespace(&client, namespace).await;
+                if !status.success() {
+                    let mut stderr_reader = BufReader::new(server.stderr.take().unwrap());
+                    let mut err_stream = String::new();
+                    stderr_reader.read_to_string(&mut err_stream).unwrap();
+                    panic!("{}", err_stream);
+                } else {
+                    panic!("server exited with status code: {}", status);
+                }
+            }
+            Ok(None) => {}
+            Err(e) => panic!("{:?}", e),
+        }
+        let child_stdout = server.stdout.take().unwrap();
+        validate_requests(child_stdout, service_url.as_str(), &mut server).await;
+        let jobs_api: Api<Job> = Api::namespaced(client.clone(), namespace);
+        let jobs = jobs_api.list(&ListParams::default()).await.unwrap();
+        assert_eq!(jobs.items.len(), 1);
+        let job_name = jobs.items[0].metadata.name.clone().unwrap();
+
+        let pods_api: Api<Pod> = Api::namespaced(client.clone(), namespace);
+        let pods = pods_api.list(&ListParams::default()).await.unwrap();
+        let pod_name = pods
+            .items
+            .iter()
+            .filter_map(|pod| {
+                if pod.metadata.name.clone().unwrap().contains(&job_name) {
+                    Some(pod.metadata.name.clone().unwrap())
+                } else {
+                    None
+                }
+            })
+            .next()
+            .unwrap();
+        assert!(pod_name.contains(&job_name));
+        delete_namespace(&client, namespace).await;
+        sleep(Duration::from_secs(10)).await; // Todo: check if we still need this sleep
```

Comment:
Why would we need a sleep at the end of the test?

Reply:
some of the tests are using the same namespace, if I dont sleep then if one test deleted the namespace and the other wants to create it immediately and somehow it wasnt deleted then we get an error from the server. however in this case its a unique namespace so we can probably remove it. ill check and update

Reply:
fixed, we don't need the sleep there